lib/shippo/api.rb

```diff
@@ -18,7 +18,21 @@ module API
     @warnings = true
 
     class << self
-      attr_accessor :base, :version, :token, :debug, :warnings
+      attr_writer :token
+      attr_accessor :base, :version, :debug, :warnings
+
+      def token
+        Thread.current[:shippo_api_token] || @token
+      end
+
+      def with_token(token)
+        old_thread_token = Thread.current[:shippo_api_token]
```

Comment:
Why is the thread storage necessary here, rather than storing the old token in a local var and changing the instance variable in the context of the block? Asking just to understand.

Reply:
I use thread variable instead of resetting `@token` to ensure thread safety in multithreaded service mode ([Puma](https://github.com/puma/puma), [Sidekiq](https://github.com/mperham/sidekiq)).

`old_thread_token` is another story - it stores previously set thread variable value. In most cases it would be `nil`, but if for some strange reason someone would nest `with_token` calls, it would ensure correct token:

```ruby
Shippo::API.with_token('token1') do
  # token1 would be used

  Shippo::API.with_token('token2') do
    # token2 would be used
  end
  
  # token1 would be used
end
```

Example using traditionally set token (stored in `@token`) + temporary tokens:

```ruby
Shippo::API.token = 'token1'
# token1 would be used

Shippo::API.with_token('token2') do
  # token2 would be used
end

# token1 would be used
```